runtime/eventring_consumer.c

```diff
@@ -0,0 +1,686 @@
+    CloseHandle(cursor->ring_handle);
+#else
+    munmap(cursor->metadata, cursor->ring_file_size_bytes);
+#endif
+    caml_stat_free(cursor->current_positions);
+    caml_stat_free(cursor);
+  }
+}
+
+eventring_error
+caml_eventring_read_poll(struct caml_eventring_cursor *cursor,
+                         void *callback_data, uintnat max_events,
+                         uintnat *events_consumed) {
+  int consumed = 0;
+  int start_domain = cursor->next_read_domain;
+  uint64_t ring_head, ring_tail;
+  int early_exit = 0;
+
+  if (!cursor->cursor_open) {
+    return E_CURSOR_NOT_OPEN;
+  }
+
+  /* this loop looks a bit odd because we're iterating from the last domain
+     that we read from on the last read_poll call and then looping around.
+     This is necessary because in the case where the consumer can't keep up
+     with message production (i.e max_events is hit each time) it ensures that
+     messages are read from all domains, rather than just the first. */
+  for (int i = 0; i < cursor->metadata->max_domains && !early_exit; i++) {
+    int domain_num = (start_domain + i) % cursor->metadata->max_domains;
+
+    struct eventring_buffer_header *eventring_buffer_header =
+        (struct eventring_buffer_header *)(
+          (char*)cursor->metadata +
+          cursor->metadata->headers_offset +
+          domain_num * cursor->metadata->ring_header_size_bytes
+        );
+
+    uint64_t *ring_ptr = (uint64_t *)((char*)cursor->metadata +
+                                      cursor->metadata->data_offset +
+                                domain_num * cursor->metadata->ring_size_bytes);
+
+    do {
+      uint64_t buf[EVENTRING_MAX_MSG_LENGTH];
+      uint64_t ring_mask, header, msg_length;
+      ring_head = atomic_load_explicit(&eventring_buffer_header->ring_head,
+                                       memory_order_acquire);
+      ring_tail = atomic_load_explicit(&eventring_buffer_header->ring_tail,
+                                       memory_order_acquire);
+
+      if (ring_head > cursor->current_positions[domain_num]) {
+        if (cursor->lost_events) {
+          cursor->lost_events(domain_num, callback_data,
+                                    ring_head -
+                                        cursor->current_positions[domain_num]);
+        }
+        cursor->current_positions[domain_num] = ring_head;
+      }
+
+      if (cursor->current_positions[domain_num] >= ring_tail) {
+        break;
+      }
+
+      ring_mask = cursor->metadata->ring_size_elements - 1;
+      header = ring_ptr[cursor->current_positions[domain_num] & ring_mask];
+      msg_length = EVENTRING_ITEM_LENGTH(header);
+
+      if (msg_length > EVENTRING_MAX_MSG_LENGTH) {
+        return E_CORRUPT_STREAM;
+      }
+
+      memcpy(buf,
+             ring_ptr + (cursor->current_positions[domain_num] & ring_mask),
+             msg_length * sizeof(uint64_t));
+
+      ring_head = atomic_load_explicit(&eventring_buffer_header->ring_head,
```

Comment:
I want to flag this memory order because I think it's wrong, but I'm not 100%. My understanding is that acquire means you can't move reads or writes before this load. Which you want. But you also don't want the memcpy on the previous line to be moved after this atomic load. So I think you need memory_order_acq_rel here, and only here. memory_order_acq_rel is an acquire and a release.

Reply:
I've been pondering this one and I'm not 100% sure either. `memory_order_acq_rel` only seems to apply to read-modify-write.

Reply:
Having sought out some experts, a seqcst memory fence after the memcpy should suffice. I've done this in https://github.com/ocaml/ocaml/pull/10964/commits/5ddc236aecc04b6f358848ea72006f0fc690b3ae